Migrate Equivocation Handling to x/evidence

alexanderbez opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

* Move `HandleDoubleSign` from `x/slashing` to the `x/evidence` module
* Move double sign params from `x/slashing` to the `x/evidence` module
* Cleanup specs
* Add params handler to `x/evidence` client logic

closes: #5296

----

- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/develop/CONTRIBUTING.md#pr-targeting))

- [x] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [x] Wrote tests
- [x] Updated relevant documentation (`docs/`)
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

fedekunze commented on x/evidence/internal/keeper/infraction.go:
```diff
@@ -0,0 +1,104 @@
+	}
+
+	// reject evidence if the double-sign is too old
+	if age > k.MaxEvidenceAge(ctx) {
```
q: shouldn't every evidence type define its own evidence age instead?

alexanderbez replied:
I don't think so because all evidence must be within the unbonding period. Perhaps certain evidence can have even smaller valid periods (e.g. 1/2 unbonding period), but I don't see any immediate use-cases or evidence types that need this.

fedekunze commented on x/evidence/internal/keeper/params.go:
```diff
@@ -0,0 +1,25 @@
+)
+
+// MaxEvidenceAge returns the maximum age for submitted evidence.
+func (k Keeper) MaxEvidenceAge(ctx sdk.Context) (res time.Duration) {
```
see question above

alexanderbez replied:
see answer :)

fedekunze commented on x/evidence/internal/types/evidence.go:
```diff
@@ -0,0 +1,104 @@
+}
+
+// ValidateBasic performs basic stateless validation checks on an Equivocation object.
+func (e Equivocation) ValidateBasic() error {
```
missing check for default time:

```go
if e.Time.IsZero() {
...
}
```

alexanderbez replied:
good catch; updated + added unit tests.

fedekunze commented on x/slashing/internal/keeper/signing_info.go:
```diff
@@ -77,6 +87,73 @@ func (k Keeper) IterateValidatorMissedBlockBitArray(ctx sdk.Context,
+		),
+	)
+
+	fraction := k.SlashFractionDoubleSign(ctx)
```
I know this is the case for double signing, but should we make the `Slash` function general enough so that other modules can use it too with their custom slashing fractions? Eg: tBTC and Peggy need to use the slashing logic too. 

If you disagree (which could be ok), what would be the recommended steps for them to implement slashing with their custom fractions for their use cases?

alexanderbez replied:
wooaaa, great catch! This was **not** my intention and is a bug! The fraction needs to be passed to this method -- making it general.

alexanderbez replied:
updated to take a fraction which the evidence module supplies.

fedekunze commented on x/slashing/internal/keeper/signing_info.go:
```diff
@@ -1,6 +1,9 @@
 package keeper
```
should we rename this file to `slash.go` or `punishment.go` as it now contains logic other than the signing info ?

alexanderbez replied:
Meh, I want to leave `signing_info.go` as-is, but I will move non-related methods to `keeper.go`.

fedekunze commented on x/evidence/module.go:
```diff
@@ -158,7 +158,9 @@ func (am AppModule) ExportGenesis(ctx sdk.Context) json.RawMessage {
 // BeginBlock executes all ABCI BeginBlock logic respective to the evidence module.
-func (AppModule) BeginBlock(_ sdk.Context, _ abci.RequestBeginBlock) {}
+func (am AppModule) BeginBlock(ctx sdk.Context, req abci.RequestBeginBlock) {
+	BeginBlocker(ctx, req, am.keeper)
```
what's the naming convention for `BeginBlock` vs `BeginBlocker`?

alexanderbez replied:
I see the two used interchangeably. `BeginBlock` is the defacto nomenclature, but we use `BeginBlocker` through the SDK in certain places. We should clean this up -- in a separate PR of course.

fedekunze commented on x/evidence/abci.go:
```diff
@@ -0,0 +1,25 @@
+	tmtypes "github.com/tendermint/tendermint/types"
+)
+
+func BeginBlocker(ctx sdk.Context, req abci.RequestBeginBlock, k Keeper) {
```
nit missing godoc. Moving the comment below here suffices

alexanderbez replied:
updated.

AdityaSripal commented on x/evidence/internal/types/evidence.go:
```diff
@@ -0,0 +1,104 @@
+	if e.Power < 1 {
+		return fmt.Errorf("invalid equivocation validator power: %d", e.Power)
+	}
+	if len(e.ConsensusAddress) == 0 {
```
```suggestion
	if e.ConsensusAddress.Empty() {
```

AdityaSripal commented on x/evidence/client/rest/query.go:
```diff
@@ -22,6 +22,11 @@ func registerQueryRoutes(cliCtx context.CLIContext, r *mux.Router) {
+	r.HandleFunc(
+		"/evidence/params",
+		queryParamsHandler(cliCtx),
+	).Methods("GET")
```
Should this be changed to:

```suggestion
	).Methods(MethodGet)
```

AdityaSripal commented on x/evidence/internal/types/params.go:
```diff
@@ -0,0 +1,60 @@
+
+	// The Double Sign Jail period ends at Max Time supported by Amino
+	// (Dec 31, 9999 - 23:59:59 GMT).
+	DoubleSignJailEndTime = time.Unix(253402300799, 0)
```
Shouldn't this be a parameter as well?

alexanderbez replied:
This is because validators that equivocate are tombstoned and as a result are technically jailed "forever" -- hence it doesn't make sense for this to be configurable. i.e. it must be the maximum time allowed. If we ever remove tombstoning, then it can be configurable.

AdityaSripal commented:
Also a heads-up, the Evidence interface got changed in #5321 to accomodate chain misbehaviour. Might require some changes to this PR but nothing substantial.

Would just require asserting that the Equivocation struct satified `ValidatorEvidence` interface rather than looser `Evidence` interface

alexanderbez commented:
@AdityaSripal @fedekunze I've updated and addressed all comments.

----

> Also a heads-up, the Evidence interface got changed in #5321 to accomodate chain misbehaviour. Might require some changes to this PR but nothing substantial.

Would just require asserting that the Equivocation struct satified `ValidatorEvidence` interface rather than looser `Evidence` interface

How so? What if we merge this PR first? I haven't looked at the PR yet, but I imagine the stricter interface contract is warranted?

AdityaSripal commented:
The new `ValidatorEvidence` interface is identical to the current `Evidence` interface.

The original `Evidence` interface was made looser in #5321 to accommodate other forms of Evidence that are not Validator-specific; namely entire chains misbehaving

alexanderbez commented:
Great! Sounds like a plan. Thanks for the review <3
